### ISSUE TITLE
Php-fpm plugin

### ISSUE THREAD

**Issue by cornerot** [influxdb/telegraf#234]
Add php-fpm stats

**Comment by sparrc** [comment]
Thanks @cornerot, any other details you can provide would be much appreciated!

**Comment by mabushey** [comment]
It looks something like this:
```
SCRIPT_NAME=/fpm-status-zwei SCRIPT_FILENAME=/fpm-status-zwei REQUEST_METHOD=GET cgi-fcgi -bind -connect 127.0.0.1:9001
X-Powered-By: PHP/5.6.13-1~dotdeb+7.1
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Cache-Control: no-cache, no-store, must-revalidate, max-age=0
Content-type: text/plain;charset=UTF-8

pool:                 www0
process manager:      dynamic
start time:           21/Dec/2013:17:01:24 -0800
start since:          56069832
accepted conn:        8987
listen queue:         0
max listen queue:     0
listen queue len:     128
idle processes:       4
active processes:     1
total processes:      5
max active processes: 7
max children reached: 0
slow requests:        123
```

It's typical to run multiple pools (I have 9001 and 9002, and yes, I know 9001 is named two). Also  `QUERY_STRING=full` can be specified:
```
# SCRIPT_NAME=/fpm-status SCRIPT_FILENAME=/fpm-status QUERY_STRING=full REQUEST_METHOD=GET cgi-fcgi -bind -connect 127.0.0.1:9002
X-Powered-By: PHP/5.6.13-1~dotdeb+7.1
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Cache-Control: no-cache, no-store, must-revalidate, max-age=0
Content-type: text/plain;charset=UTF-8

pool:                 www1
process manager:      dynamic
start time:           21/Dec/2013:17:01:24 -0800
start since:          56070033
accepted conn:        8983
listen queue:         0
max listen queue:     0
listen queue len:     128
idle processes:       4
active processes:     1
total processes:      5
max active processes: 8
max children reached: 0
slow requests:        104

************************
pid:                  19031
state:                Running
start time:           01/Oct/2015:14:52:58 -0700
start since:          7739
requests:             49
request duration:     118
request method:       GET
request URI:          /fpm-status?full
content length:       0
user:                 -
script:               -
last request cpu:     0.00
last request memory:  0

************************
pid:                  25444
state:                Idle
start time:           01/Oct/2015:15:31:12 -0700
start since:          5445
requests:             37
request duration:     749848
request method:       GET
request URI:          /index.php
content length:       0
user:                 -
script:               /var/www/html/www.example.com/index.php
last request cpu:     81.35
last request memory:  101187584

************************
pid:                  6633
state:                Idle
start time:           01/Oct/2015:14:40:03 -0700
start since:          8514
requests:             61
request duration:     5335317
request method:       GET
request URI:          /index.php?token=hf5HyAqt5EXSvBc3LuQWrhsQFK1Iyqnz6_FIJEz-L64&destination=node/84444
content length:       0
user:                 -
script:               /var/www/html/example.com/index.php
last request cpu:     90.53
last request memory:  192675840

************************
pid:                  6827
state:                Idle
start time:           01/Oct/2015:14:40:28 -0700
start since:          8489
requests:             53
request duration:     177470
request method:       POST
request URI:          /index.php
content length:       8996
user:                 -
script:               /var/www/html/example.com/index.php
last request cpu:     95.79
last request memory:  36175872

************************
pid:                  29064
state:                Idle
start time:           01/Oct/2015:15:32:53 -0700
start since:          5344
requests:             25
request duration:     815690
request method:       GET
request URI:          /index.php
content length:       0
user:                 -
script:               /var/www/html/example.com/index.php
last request cpu:     78.46
last request memory:  100925440
```

In addition to the status, there is a ping:
```
# SCRIPT_NAME=/ping-zwei SCRIPT_FILENAME=/ping-zwei REQUEST_METHOD=GET cgi-fcgi -bind -connect 127.0.0.1:9001
X-Powered-By: PHP/5.6.13-1~dotdeb+7.1
Content-type: text/plain;charset=UTF-8
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Cache-Control: no-cache, no-store, must-revalidate, max-age=0

pong
```


This is configured in the files `/etc/php5/fpm/pool.d/www0.conf` and `/etc/php5/fpm/pool.d/www1.conf`. From www0.conf:
```
27 listen = 127.0.0.1:9001
139 pm.status_path = /fpm-status-zwei
151 ping.path = /ping-zwei
155 ; Default Value: pong
156 ;ping.response = pong
```

**Comment by sparrc** [comment]
@mabushey so this information is available on `localhost:9001/fpm-status-zwei` and `localhost:9001/ping-zwei`? Does the URL always have the `-zwei` added or is that something that you've configured?

**Comment by mabushey** [comment]
These are specified in the conf files located in `/etc/php5/fpm/pool.d/` (`www0` and `www1` in my case). I think you missed that part of my last post:

The listen directive (line 27) determines the IP/port binding.
The `pm.status_path` (line 139) and `ping.path` (line 151) control the URL path.

**Comment by kureikain** [comment]
I'm writing PHP code so probably I'm more familiar with this stuff. If no one wanna take it, can you assign to me to prevent other dev work on it? It's similar to my HAProxy plugin, where I can request an end-point, just different parser to grab data point.

**Comment by sparrc** [comment]
@kureikain I'm not able to assign to someone outside of the InfluxDB organization, but don't worry, nobody else has claimed this so work away! Thank you! :+1:

**Comment by kureikain** [comment]
Cool, will adjust HAProxy plugin to parse FPM stat output.

**Comment by aderumier** [comment]
Hi,
listen can also be an unix socket, it could be great to be able to read it too.

**Comment by kureikain** [comment]
@aderumier I just had the stat via HTTP first. WIll try to get the unix socket working, but prlly not on first iteration.

**Comment by cornerot** [comment]
This draft. Pool-request and test I will do later.
```go
package phpfpm

import (
	"bufio"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/influxdb/telegraf/plugins"
)

type Phpfpm struct {
	Urls []string
}

var sampleConfig = `
	# An array of Phpfpm status URI to gather stats.
	urls = ["http://localhost/status-localhost"]
`

func (n *Phpfpm) SampleConfig() string {
	return sampleConfig
}

func (n *Phpfpm) Description() string {
	return "Read Phpfpm's status information"
}

func (n *Phpfpm) Gather(acc plugins.Accumulator) error {
	var wg sync.WaitGroup
	var outerr error

	for _, u := range n.Urls {
		addr, err := url.Parse(u)
		if err != nil {
			return fmt.Errorf("Unable to parse address '%s': %s", u, err)
		}

		wg.Add(1)
		go func(addr *url.URL) {
			defer wg.Done()
			outerr = n.gatherUrl(addr, acc)
		}(addr)
	}

	wg.Wait()

	return outerr
}

var tr = &http.Transport{
	ResponseHeaderTimeout: time.Duration(3 * time.Second),
}

var client = &http.Client{Transport: tr}

func (n *Phpfpm) gatherUrl(addr *url.URL, acc plugins.Accumulator) error {
	resp, err := client.Get(addr.String())
	if err != nil {
		return fmt.Errorf("error making HTTP request to %s: %s", addr.String(), err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("%s returned HTTP status %s", addr.String(), resp.Status)
	}

	tags := getTags(addr)

	sc := bufio.NewScanner(resp.Body)
	for sc.Scan() {
		line := sc.Text()
		parts := strings.SplitN(line, ":", 2)
		key, part := strings.Replace(parts[0], " ", "_", -1), strings.TrimSpace(parts[1])
		switch key {
		case "process_manager":
			tags["process_manager"] = part
		case "pool":
			tags["pool"] = part
		This draft.default:
			value, err := strconv.ParseUint(part, 10, 64)
			if err != nil {
				continue
			}
			acc.Add(key, value, tags)
		}
	}

	return nil
}

// Get tag(s) for the phpfpm plugin
func getTags(addr *url.URL) map[string]string {
	h := addr.Host
	host, port, err := net.SplitHostPort(h)
	if err != nil {
		host = addr.Host
		if addr.Scheme == "http" {
			port = "80"
		} else if addr.Scheme == "https" {
			port = "443"
		} else {
			port = ""
		}
	}
	return map[string]string{"server": host, "port": port}
}

func init() {
	plugins.Add("phpfpm", func() plugins.Plugin {
		return &Phpfpm{}
	})
}
```

**Comment by kureikain** [comment]
PR is added https://github.com/influxdb/telegraf/pull/255

**Comment by cornerot** [comment]
ok :+1:

**Comment by kureikain** [comment]
@aderumier socket is also supported. Could you give it a try once it's merged.

**Comment by sparrc** [comment]
closed by #255
